// @jitl/notion-api/src/example/recipeCMS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const __1 = require("..");
const content_management_system_1 = require("../lib/content-management-system");
const exampleHelpers_1 = require("./exampleHelpers");
const DEBUG_EXAMPLE = __1.DEBUG.extend('example');
(0, exampleHelpers_1.runExample)(module, 'Recipe CMS', async ({ notion }) => {
    const Recipes = new content_management_system_1.CMS({
        database_id: 'a3aa29a6b2f242d1b4cf86fb578a5eea',
        notion,
        slug: undefined,
        visible: true,
        getFrontmatter: () => ({}),
        schema: {},
        cache: {
            directory: path.join(__dirname, './cache'),
        },
        assets: {
            directory: path.join(__dirname, './assets'),
            downloadExternalAssets: true,
        },
    });
    if (!process.env.DEBUG) {
        process.env.DEBUG = '@jitl/notion-api:*';
        console.log('DEBUG', process.env.DEBUG);
    }
    // Download and cache all pages in the Recipes database, and their assets.
    for await (const recipe of Recipes.query()) {
        const s = (0, __1.richTextAsPlainText)(recipe.frontmatter.title);
        DEBUG_EXAMPLE('page "%s" with children %d', s, recipe.content.children.length);
        await Recipes.downloadAssets(recipe);
    }
});
